tests/jdom/fastforwardtester.ts

```diff
@@ -0,0 +1,187 @@
+import { JDBus } from "../../src/jdom/bus"
+import { DEVICE_ANNOUNCE } from "../../src/jdom/constants"
+import { JDDevice } from "../../src/jdom/device"
+import JDServiceProvider from "../../src/jdom/serviceprovider"
+import JDServiceServer from "../../src/jdom/serviceserver"
+import { assert } from "../../src/jdom/utils"
+import {
+    DebugConsoleUi,
+    SynchronizationTimingOptions,
+    TestDriver,
+    TestDriverInterface,
+    TesterEvent,
+    WaitTimingOptions,
+} from "../../src/tstester/base"
+import { BusTester } from "../../src/tstester/testwrappers"
+import { loadSpecifications } from "../testutils"
+import { FastForwardScheduler } from "./scheduler"
+import { CreatedServerService } from "./tester"
+
+// Set equals is not a built-in operation.
+function setEquals<T>(set1: Set<T>, set2: Set<T>): boolean {
+    if (set1.size != set2.size) {
+        return false
+    }
+    set1.forEach(value => {
+        if (!set2.has(value)) {
+            return false
+        }
+    })
+    return true
+}
+
+// A bus test wrapper that uses the fast-forward bus and provides test helper functions
+// to spin up virtual devices.
+//
+// This additionally wraps the TestDriver interfaces with logic needed to advance simulator time.
+// Test code should not create a new TestDriver interface, and instead use the functions exposed
+// by this class.
+export class FastForwardTester
+    extends BusTester
+    implements TestDriverInterface
+{
+    static makeTest(test: (bus: FastForwardTester) => Promise<void>) {
+        return async () => {
+            await this.withTestBus(test)
+        }
+    }
+
+    // Wrapper that provides bus construction, initializaiton, and teardown
+    static async withTestBus(test: (bus: FastForwardTester) => Promise<void>) {
```

Comment:
not a fan of static function, can we move those to top level functions?

Reply:
Done.